ls1intum/Artemis
# `Programming exercises:` Add the option to record the testwise coverage for Java gradle programming exercises

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab). (Should not be possible since Gradle projects are not available on the GitLab/Jenkins-Setup)

### Motivation and Context
These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
In order to create code hints for behavioral test cases, we will need to record the testwise coverage for the solution repository.
The testwise coverage data is collected by the [Teamscale JaCoCo agent](https://github.com/cqse/teamscale-jacoco-agent). The generated coverage data follows [this data model](https://docs.teamscale.com/howto/providing-testwise-coverage/#sample-file-for-testwise-coverage).

### Description
Within these changes, an option for Java-Gradle programming exercises with non-sequential is added. The creation of the build plans is adjusted, the creation of the test-repository build.gradle.

In addition, this PR contains changes regarding the database. This diagram describes the newly introduced domain models:
![ClassDiagram (11)](https://user-images.githubusercontent.com/73833780/161739573-2c2ff0e4-3247-4d82-ae57-e05404205300.png)


Note that this is just the first PR on this new feature. These points will be done in follow-up PRs:
- Add support for sequential test runs
- Improve the UI/UX of the modal
- Display more statisticcs on coverage on exercise details page
- Add support for the GitLab/Jenkins-Setup. This is currently not possible because Gradle projects are not available on this setup.
- Generate the coverage diff between template and solution repository (will be calculated with git-diff and solution-coverage)
- Display an information to the instructor if the same lines of code are executed multiple times across multiple test cases

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Test-Server with the Atlassian-Prelive-Setup (TS1, TS5)

1. Log in to Artemis
2. Navigate to Course Administration
3. Generate a new Programming Exercise with the following configuration
- Project Type: Plain - Gradle or Gradle - Gradle, this does not matter
- Option selected: "Record Testwise Coverage"
- SCA: enabled
4. Click "Generate" and wait until the solution build has finished
5. Click the button "Show Testwise Coverage" and play a little bit around. Look into the different files, select and se-select the different test cases for which the covered lines should be displayed. Check that the covered lines are marked yellow and that the covered lines ratio updates accordingly to the test case selection.
6. Verify that the value at the bottom of the programming exercise details page for "Ratio Test-Covered Lines" is correct (or seems reasonable).
7. Re-run the build plan and verify that everything works as expected.
8. Push to the solution and template repository and repeat steps 5 - 7.
9. Delete the exercise and verify that this works correctly.
10. Create a Maven exercise and verify that this still works as expected.

**Note**: The data which is used to generate the coverage is created in the build of the solution submission. If you open the solution build plan, click on 'Artifacts' and download the file 'testwiseCoverageReport' you can verify the result of the data displayed in Artemis with the recorded data (they should be the same).

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| CoverageReportRepository.java | 100% | 100% |
| TestwiseCoverageService.java | 80% | 94.3% |
| CoverageReportResource.java | n.a. | 100% |
| testwise-coverage-file.component.ts | 70.6% | 95.0% |
| testwise-coverage-report.component.ts | 76.5% | 96.7% |

### Screenshots
The option on generation:
![Bildschirmfoto 2022-04-01 um 19 04 38](https://user-images.githubusercontent.com/73833780/161310114-f58cc425-ba8d-4c62-9fa7-6d4197d59b43.png)

The new modal:
![Bildschirmfoto 2022-04-03 um 16 40 31](https://user-images.githubusercontent.com/73833780/161433366-96378996-b5ac-4a85-8833-926ef42b8ed2.png)

![Bildschirmfoto 2022-04-03 um 16 40 45](https://user-images.githubusercontent.com/73833780/161433367-44305e68-c33c-49c0-aa5a-5210f7ebf8d0.png)
